_posts/2020-12-02-pipelined-region-sheduling.md

```diff
@@ -0,0 +1,284 @@
+Each _operator_ has a certain number of _subtasks_ executed in parallel. The _subtask_ is the actual execution unit in Flink.
+Each subtask can consume user records from other subtasks (inputs), process them and produce records for further consumption by other _subtasks_ (outputs) down the stream.
+There are _source subtasks_ without inputs and _sink subtasks_ without outputs. Hence, the _subtasks_ form the nodes of the
+_[ExecutionGraph](https://ci.apache.org/projects/flink/flink-docs-release-1.11/internals/job_scheduling.html#jobmanager-data-structures)_.
+
+
+### Intermediate results
+
+There are also two major data-exchange types to produce and consume results by _operators_ and their _subtasks_: _pipelined_ and _blocking_.
+They are basically types of the edges in the _ExecutionGraph_.
+
+The _pipelined_ result can be consumed record by record. It means that the consumer can already run once the first result records have been produced.
+The _pipelined_ result can be a never ending output of records, e.g. in case of a _streaming job_.
+
+The _blocking_ result can be consumed only when its _production_ is done. Hence, the _blocking_ result is always finite
+and the consumer of the _blocking_ result can run only when the producer has finished its execution.
+
+### Slots and resources
+
+Each _[TaskManager](https://ci.apache.org/projects/flink/flink-docs-release-1.11/concepts/flink-architecture.html#anatomy-of-a-flink-cluster)_
+instance has a certain number of virtual _[slots](https://ci.apache.org/projects/flink/flink-docs-release-1.11/concepts/flink-architecture.html#task-slots-and-resources)_.
+Each _slot_ represents a certain part of the _TaskManager’s physical resources_ to run the operator _subtasks_.
+Each _subtask_ is deployed into a _slot_ of the _TaskManager_.
+Each _slot_ can run multiple _[subtasks](https://ci.apache.org/projects/flink/flink-docs-release-1.11/internals/job_scheduling.html#scheduling)_
+at the same time from different _operators_, usually [chained](https://ci.apache.org/projects/flink/flink-docs-release-1.11/concepts/flink-architecture.html#tasks-and-operator-chains) together.
+
+### Scheduling strategy
+
+_[Scheduling](https://ci.apache.org/projects/flink/flink-docs-release-1.11/internals/job_scheduling.html#scheduling)_
+in Flink is a process of searching for and allocating appropriate resources (_slots_) from the _TaskManagers_ to run the _subtasks_ and produce results.
+The _scheduling strategy_ reacts on scheduling events (like start job, _subtask_ failed or finished etc) to decide which _subtask_ to deploy next.
+
+For instance, it does not make sense to schedule _subtasks_ which inputs are not ready to consume yet to avoid waste of resources.
+Another example is to schedule _subtasks_, which are connected with _pipelined_ edges, together to avoid deadlocks caused by backpressure.
```

Comment:
```suggestion
For instance, it does not make sense to schedule _subtasks_ whose inputs are not ready to consume yet to avoid wasting resources.
Another example is to schedule _subtasks_ which are connected with _pipelined_ edges together, to avoid deadlocks caused by backpressure.
```